src/radical/pilot/resource_config.py

```diff
@@ -194,3 +201,508 @@ class ResourceConfig(ru.TypedDict):
+                self.gpus[ro.index].occupation += ro.occupation
+
+            self.lfs -= lfs
+            self.mem -= mem
+
+          # print('      => %s' % self)
+          # print('allocate %s' % slot)
+          # print()
+
+
+    # --------------------------------------------------------------------------
+    #
+    def deallocate_slot(self, slot : 'Slot') -> None:
+
+        with self.__lock:
+
+          # print('release  %s' % slot)
+          # print('     ->  %s' % self)
+            for ro in slot.cores:
+                self.cores[ro.index].occupation -= ro.occupation
+              # assert self.cores[ro.index].occupation >= 0.0, \
+              #         'invalid core release: %s' % self
+
+            for ro in slot.gpus:
+                self.gpus[ro.index].occupation -= ro.occupation
+              # assert self.gpus[ro.index].occupation >= 0.0, \
+              #         'invalid gpu release: %s' % self
+
+            self.lfs += slot.lfs
+            self.mem += slot.mem
+
+          # print('     =>  %s' % self)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def find_slot(self, rr: RankRequirements) -> Optional[Slot]:
+
+        with self.__lock:
+
+            cores = list()
+            gpus  = list()
+
+            # NOTE: the current mechanism will never use the same core or gpu
+            #       multiple times for the created slot, even if the respective
+            #       occupation would allow for it.
+            if rr.n_cores:
+                for ro in self.cores:
+                    if ro.occupation is DOWN:
+                        continue
+                    if rr.core_occupation <= 1 - ro.occupation:
+                        cores.append(_RO(index=ro.index,
+                                         occupation=rr.core_occupation))
+                    if len(cores) == rr.n_cores:
+                        break
+
+                if len(cores) < rr.n_cores:
+                    return None
+
+            if rr.n_gpus:
+                for ro in self.gpus:
+                    if ro.occupation is DOWN:
+                        continue
+                    if rr.gpu_occupation <= 1 - ro.occupation:
+                        gpus.append(_RO(index=ro.index,
+                                        occupation=rr.gpu_occupation))
+                    if len(gpus) == rr.n_gpus:
+                        break
+
+                if len(gpus) < rr.n_gpus:
+                    return None
+
+            if rr.lfs and self.lfs < rr.lfs: return None
+            if rr.mem and self.mem < rr.mem: return None
+
+            slot = Slot(cores=cores, gpus=gpus, lfs=rr.lfs, mem=rr.mem,
+                        node_index=self.index, node_name=self.name)
+            self.allocate_slot(slot, _check=False)
+
+            return slot
+
+
+# ------------------------------------------------------------------------------
+#
+class NodeList(ru.TypedDict):

```

Comment:
This class would be even more useful if it had methods to query the number of available nodes, cores, gpus, etc.

During the heartbeat of the executor, I would like to be able to call something like the following so I can decide whether to look up the requirements for the remaining tasks to determine whether there are any available resources that could potentially satisfy task requirements.

```
if NodeList.num_cores():
    return # No point in finding task requirements if there are no free cores
```

Reply:
It is not strictly necessary to check if there are any free cores. The utility comes in that checking how much resources a task requires requires querying the database, which which many other types of operations also need to do. Checking if there are any free cores is an operation that is local to to executor and can save a potentially large number of DB queries. If you are reluctant to add interface to your class I totally understand. I just wanted to raise the point that it could be a useful feature which in particular prevents potential caller from needing to understand the internals of the class in terms of the accounting. Perhaps just a bool `slots_available()` or similar would be sufficient, but I thought getting the number of free cores might be even more useful.

Reply:
> First, there is no database query involved

Sorry for not being clear. The DB queries I want to avoid are in airflow, not rp. I am aware that a class which is basically just a fancy dict does not have any significant performance overhead. I have not done any testing to see what the limits of e.g. PostgreSQL are in an HPC context but my intuition says fewer queries are better.

I will go ahead and use `if not nl.find_slots({'n_cores': 1}, n_slots=1):`.